third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -0,0 +1,1453 @@
+                    format!(
+                        "cannot immutable borrow {} since other mutable references exist",
+                        msg_for_source()
+                    ),
+                    "immutable borrow attempted here",
+                    diags.into_iter(),
+                )
+            }
+        }
+        state.add_edge(label, edge)
+    }
+
+    fn error_with_hints(
+        &self,
+        loc: &Loc,
+        msg: impl AsRef<str>,
+        primary: impl AsRef<str>,
+        hints: impl Iterator<Item = (Loc, String)>,
+    ) {
+        self.target.global_env().diag_with_primary_and_labels(
+            Severity::Error,
+            loc,
+            msg.as_ref(),
+            primary.as_ref(),
+            hints.collect(),
+        )
+    }
+
+    fn borrow_info(
+        &self,
+        state: &LifetimeState,
+        label: LifetimeLabel,
+        only_mut: bool,
+        alive: &LiveVarInfoAtCodeOffset,
+    ) -> Vec<(Loc, String)> {
+        let primary_edges = state
+            .effective_descendants(label)
+            .iter()
+            .cloned()
+            .collect::<BTreeSet<_>>();
+        let mut secondary_edges = BTreeSet::new();
+        for (_, edge) in primary_edges.iter() {
+            if let MemoryLocation::Local(temp) = &state.graph[&edge.target].location {
+                // Include secondary edge only if it is a local which has gone out of
+                // scope. In this case, the user may wonder why they get the error
+                // message, and showing derived references helps. Otherwise derived
+                // references may be noisy.
+                if !alive.before.contains_key(temp) {
+                    secondary_edges.extend(state.effective_descendants(edge.target))
+                }
+            }
+        }
+        primary_edges
+            .into_iter()
+            .filter_map(|(_, e)| self.borrow_edge_info("previous ", state, only_mut, e))
+            .chain(
+                secondary_edges
+                    .into_iter()
+                    .filter_map(|(_, e)| self.borrow_edge_info("used by ", state, only_mut, e)),
+            )
+            .collect()
+    }
+
+    fn borrow_edge_info(
+        &self,
+        prefix: &str,
+        _state: &LifetimeState,
+        only_mut: bool,
+        e: &BorrowEdge,
+    ) -> Option<(Loc, String)> {
+        if e.is_mut || !only_mut {
+            if let Some(loc) = &e.loc {
+                use BorrowEdgeKind::*;
+                let mut_prefix = if e.is_mut { "mutable " } else { "" };
+                return Some((
+                    loc.clone(),
+                    format!("{}{}{}", prefix, mut_prefix, match e.kind {
+                        BorrowLocal => "local borrow",
+                        BorrowGlobal => "global borrow",
+                        BorrowField(..) => "field borrow",
+                        BorrowIndex => "index borrow",
```

Comment:
Yes, it is actually not needed. at least for now. Vector indexing is just a `native fun _[_]<T>(v: &vector<T>, i: u64): &T`, which by default creates a `Call` borrow edge from the vector to the element. If we would start reason about the actual index value, this would be a different story, but this is hard (and in general undecidable).